File: v/CmdQuote.java
package v;

import java.util.*;

/* The world is a quote
 * */

public class CmdQuote implements Quote {
    /* The quote contains a Token stream
     * */

    // Our symbols to evaluate.
    TokenStream _tokens = null;

    public CmdQuote(TokenStream tokens) {
        _tokens = tokens;
    }

    private boolean cando(VStack stack) {
        if (stack.empty())
            return false;
        if (stack.peek().type == Type.TSymbol)
            return true;
        return false;
    }

    /* Evaluate the current stack
     * logic:
     * 		Check if our stack has any applyabl tokens (symbols).
     * if we have, apply it on the stack. if not, get next from the
     * tokenstream and repeat the procedure. If it is a compound '['
     * then push the entire quote rather than the first one.
     * */

    public void eval(VFrame scope) {
        VStack stack = scope.stack();
        Iterator<Term> stream = _tokens.iterator();
        while(stream.hasNext()) {
            stack.push(stream.next());
            if (cando(stack))
                dofunction(scope);
        }
    }

    public void dofunction(VFrame scope) {
        VStack st = scope.stack();
        // pop the first token in the stack
        Token sym = st.pop();
        Quote q = scope.lookup(sym.value());
        if (q == null)
            throw new VException("err:undef_symbol",sym, sym.value());
        // Invoke the quote on our quote by passing us as the parent.
        try {
            q.eval(scope.child());
        } catch (VException e) {
            e.addLine(sym.value());
            throw e;
        }
    }

    public TokenStream tokens() {
        return _tokens;
    }

    public String toString() {
        StringBuffer sb = new StringBuffer();
        sb.append("[");
        Iterator<Term> i = _tokens.iterator();
        while(i.hasNext()) {
            sb.append(i.next().value());
            if (i.hasNext())
                sb.append(" ");
        }
        sb.append("]");
        return sb.toString();
    }

    HashMap<String, Object> _store = new HashMap<String, Object>();
    public HashMap<String, Object> store() {
        return _store;
    }
}

File: csrc/cmdquote.cpp
#include <sstream>
#include "cmdquote.h"
#include "vframe.h"
#include "vstack.h"
#include "cmd.h"
#include "quotestream.h"
#include "lexstream.h"
#include "buffcharstream.h"
#include "quoteiterator.h"
#include "vexception.h"
#include "prologue.h"

CmdQuote::CmdQuote(TokenStream* tokens):_tokens(tokens),_val(0) {
}

void CmdQuote::eval(VFrame* scope) {
    VStack_ stack = scope->stack();
    TokenIterator_ stream = _tokens->iterator();
    while(stream->hasNext()) {
        stack->push(stream->next());
        if (cando(stack))
            dofunction(scope);
    }
}

TokenStream* CmdQuote::tokens() {
    return _tokens;
}

void CmdQuote::dofunction(VFrame* scope) {
    Token_ sym = scope->stack()->pop();
    Quote_ q = scope->lookup(sym->svalue());
    if (!q)
        throw VException("err:undef_symbol", sym, sym->value());
    try {
        VFrame_ s = scope->child(); // kludge for keeping a ref.
        q->eval(s);
        gc();
    } catch (VException& e) {
        e.addLine(sym->value());
        throw e;
    }
}

bool CmdQuote::cando(VStack* stack) {
    if (stack->empty())
        return false;
    if (stack->peek()->type() == TSymbol)
        return true;
    return false;
}

char* CmdQuote::to_s() {
    if (!_val) {
        std::ostringstream outs;
        outs << '[';
        TokenIterator_ i = _tokens->iterator();
        while(i->hasNext()) {
            outs << i->next()->value();
            if (i->hasNext())
                outs << ' ';
        }
        outs << ']';
        _val = dup_str(outs.str().c_str());
    }
    return _val;
}

Quote* CmdQuote::getdef(char* buf) {
    return new (collect) CmdQuote(
            new (collect) LexStream(new (collect) BuffCharStream(buf)));
}
